core/trino-main/src/main/java/io/trino/operator/window/pattern/MeasureComputation.java

```diff
@@ -65,43 +64,33 @@ public Type getType()
         return type;
     }
 
-    // TODO This method allocates an intermediate block and passes it as the input to the pre-compiled expression.
-    //  Instead, the expression should be compiled directly against the row navigations.
-    //  The same applies to LabelEvaluator.Evaluation.test()
-    public Block compute(int currentRow, ArrayView matchedLabels, int partitionStart, int searchStart, int searchEnd, int patternStart, long matchNumber, WindowIndex windowIndex)
+    public Block compute(int currentRow, ArrayView matchedLabels, MatchAggregation[] aggregations, int partitionStart, int searchStart, int searchEnd, int patternStart, long matchNumber, FeedablePagesWindowIndex windowIndex)
     {
-        // get values at appropriate positions and prepare input for the projection as an array of single-value blocks
+        return compute(currentRow, matchedLabels, aggregations, partitionStart, searchStart, searchEnd, patternStart, matchNumber, windowIndex, projection, expectedLayout, labelNames, session);
+    }
+
+    public Block computeEmpty(long matchNumber, MatchAggregation[] aggregations)
```

Comment:
The aggregations are indeed all the time the same instances for the instance of `MeasureComputation`, and theoretically could be passed at construction time. They are held by the `PatternRecognitionPartition`, because:
- the `PatternRecognitionPartition` knows when the aggregations need to be reset,
- the aggregations array has all the aggregations corresponding to different `MeasureComputation`s, so it's more naturally positioned in the enclosing structure.

Reply:
In that case, we should pass them at construction time instead of for each invocation (it doesn't mean that `PatternRecognitionPartition` can't also hold on to them if necessary)